pkg/envoy/ads/stream.go

```diff
@@ -47,7 +48,8 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 	// This is the Envoy proxy that just connected to the control plane.
 	// NOTE: This is step 1 of the registration. At this point we do not yet have context on the Pod.
 	//       Details on which Pod this Envoy is fronting will arrive via xDS in the NODE_ID string.
-	proxy := envoy.NewProxy(kind, uuid, si, utils.GetIPFromContext(server.Context()))
+	streamID := atomic.AddInt64(&s.nextStreamID, 1)
+	proxy := envoy.NewProxy(kind, uuid, si, utils.GetIPFromContext(server.Context()), streamID)
```

Comment:
Are we creating a proxy instance per gRPC stream? I am wondering about the impact on the control plane proxy bookkeeping workflow if that's the case.

Reply:
it is, but under the assumption that the prior stream is about to disconnect (but there are race conditions here).

I think more work will be pushed to the proxy registry (as currently we have some bookeeping happening in each per proxy stream).

With the proxy registry there are controls around querying for/listing by proxy uuid to return the proxy with the highest stream ID.

Reply:
also it should be on the order of milliseconds that there are multiple proxies connected with the same UUID

Reply:
But multiple gRPC streams are allowed per connection. Wouldn't it make sense to keep track of proxies (envoy.Proxy instance) per UUID? Creating a proxy instance per stream would mean we are tracking individual gRPC streams and not unique proxies. 

How are we certain that multiple streams can only momentarily exist per proxy?

Reply:
you're right there's no guarantee, other than knowing the envoy sidecar won't open multiple connections.

But we don't have that guarantee currently anyways. In fact, there's currently a bug where it's possible to delete the proxy from the registry after it connects.

Maintaing a registry for each connection makes sense IMO, sense it keeps track of what's connected. However the API of the registry will only return the newest connection

Reply:
I agree maintaining a proxy per connection makes sense, but connection != stream. So I am still wondering why to keep track of a proxy instance per stream, which could result in incorrect bookkeeping of unique proxies.

Reply:
maybe we're getting terminologies crossed? `Stream` refers to the grpc Stream, `StreamAggregatedResources`. We use the callback `OnStreamOpen`, which is triggered when the sidecar makes a new request to `StreamAggregatedResources`

Are you saying that a single proxy may have multiple calls to StreamAggregatedResources ongoing simultaneously?

Reply:
Though that should not happen based on observation, I don't see it being explicitly stated anywhere in the Envoy proto. Hence, I am wondering if we should be careful about bookkeeping per stream vs per proxy connection. I haven't seen this happen with the current XDS state machine, but not sure if the snapshot cache behaves differently in this regard.

Reply:
gotcha. So this implementation is "more correct" than the current implementation, in that we avoid a race condition of deleting recently joined proxies.

How we store the proxy on the registry cache is really an implementation detail IMO, and since we filter by the most recent StreamID, to an observer there's no knowledge of multiple streams.